### ISSUE TITLE
[naming-convention] Feature request: Support modifier 'literal' for variable

### ISSUE THREAD

**Issue by maple5233** [typescript-eslint/typescript-eslint#4866]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**
I want to specifying naming conventions for numeric and string literal constants which are defined with `const`.
Like this:
```ts
const BASE_URL  = 'http://foo.com' // good
const baseURL = 'http://foo.com' // bad
const MAX_NUMBER = 100 // good
const maxNumber = 100 // bad
const object = { a: 1 } // good
const a = object.a // also good, even if 'a' is a number, `object.a` can be changed to a value other than 1

```
And I have try this config:
```JS
module.exports ={
  rules: {
    '@typescript-eslint/naming-convention': [
      'error',
      {
        selector: ['variable'],
        format: ['UPPER_CASE', 'PascalCase', 'camelCase'],
        modifiers: ['const', 'destructured'],
        types: ['string', 'number'],
        leadingUnderscore: 'forbid',
        trailingUnderscore: 'forbid',
      },
      {
        selector: ['variable'],
        format: ['UPPER_CASE'],
        modifiers: ['const'],
        types: ['string', 'number'],
        leadingUnderscore: 'forbid',
        trailingUnderscore: 'forbid',
      },
    ],
  },
}


```
My test code:

```TS
/*
 * ESLint: Variable name `a` must match one of the following formats: UPPER_CASE(@typescript-eslint/naming-convention)
 * ↑ works well
 */
const a = '1'

const object = { b: '2', c: '3' }
/**
 * no eslint error, works well too
 */
const { b } = object

/**
 * ESLint: Variable name `c` must match one of the following formats: UPPER_CASE(@typescript-eslint/naming-convention
 * ↑ unexpected behavior
 */
const c = object.b

```

As the comments to the above code show, the config do not work well when we assign a property to a variable.
So, I guess shouldn't a modifier 'literal' be added for this case?

The conifg will be just like this:
```js
{
    selector: ['variable'],
    format: ['UPPER_CASE'],
    modifiers: ['const', 'literal'],
    types: ['string', 'number'],
    leadingUnderscore: 'forbid',
    trailingUnderscore: 'forbid',
},
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.12.1` |
| `@typescript-eslint/parser`        | `5.12.1` |
| `TypeScript`                       | `4.6.2` |
| `ESLint`                           | `8.10.0` |
| `node`                             | `14.18.1` |

**Comment by bradzacher** [comment]
Based solely on your current description - this doesn't seem like a broadly applicable feature request. I'd need further justification for us to consider this.

Why exactly do you specifically want variables with literal values to be named differently?

**Comment by maple5233** [comment]
> Based solely on your current description - this doesn't seem like a broadly applicable feature request. I'd need further justification for us to consider this. We aren't in the business of building and maintaining rules that cover every single use case - just enough to satisfy the majority of users.
> 
> Why exactly do you specifically want variables with literal values to be named differently?

Thanks for reply.

For literal variables defined with `const`, they can't be assigned. However,  if the developers read them elsewhere in the code, the fact that these variables can't be assigned is not obvious before they jump to read the line where these variables were defined.

As I understand, naming convention helper us to read code more conveniently. And, the industry has the habit of using UPPER_CASE to represent constants, so I hope to force these variables to use UPPER_CASE, so that when reading,  developer can directly know that these variables cannot be assigned.

**Comment by bradzacher** [comment]
Definitely - I get why you want to mark global constants to be named differently.

But why do you want ***LITERAL*** values to be treated differently to others.

What's the difference between `const x = 1;` and `const y = x;`?

**Comment by maple5233** [comment]
> Definitely - I get why you want to mark global constants to be named differently.
> 
> But why do you want _**LITERAL**_ values to be treated differently to others.
> 
> What's the difference between `const x = 1;` and `const y = x;`?

Usually, we can know in advance what their values ​​are for these values ​​that do not change. That's why we can define them literally. So we will write `const PREFIX_OF_NAME = 'prefix_'`.
For what you said `const y = x`, maybe we should use `x` directly instead of `y`. There is only one case that may be difficult to identify, that is the calculation result between literals,  like `const y = X + Z`, for this case, I haven't thought of a better way to identify it grammatically.
